### PR TITLE
feat: abstract chain + addchain scaffolder

### LINKED ISSUE #11980: EVM chain scaffolder
## tl;dr

Build a generator script that handles the mechanical ~80% of adding a second-class EVM chain, so Claude (or any agent) only needs to think about the ~20% that genuinely requires research and judgment.

Today, adding a chain touches ~44 integration points across ~35 files. The [contract](https://github.com/shapeshift/web/blob/develop/.claude/contracts/second-class-evm-chain.md) captures every gotcha. But most of those integration points are purely mechanical — append to array, add switch case, clone if-statement, create file from template. No brain needed.

**Philosophy: don't use a brain until you need a brain.**

## the approach: marker-based generation

No AST parsing (jscodeshift/ts-morph). Instead:

1. **~60 marker comments** placed at insertion points across the codebase (`// @chain-gen:<KEY>`)
2. **A simple generator** (`scripts/addChain/index.ts`) that reads a chain config JSON, finds markers, inserts templated code before them
3. **6 file templates** for new files (chain adapter, plugin, asset gen script, CSP header, tx status util)
4. **Claude handles the 20%** — swapper research, stablecoin address lookups, edge case detection, verification

### why markers over AST transforms

| | Markers | jscodeshift |
|---|---|---|
| Implementation | ~200 LOC regex engine | ~2000 LOC across 25 codemods |
| Works on .env/.d.ts | yes | no |
| Survives refactors | marker moves with code | breaks if variable renamed |
| Fails loudly | "marker not found" error | silent wrong insertion |
| Self-documenting | markers tell devs "this is extensible" | codemods are invisible |
| Maintenance | near zero | constant babysitting |

## chain config schema

Single JSON file per chain with everything the generator needs:

```json
{
  "name": "Bob",
  "camelName": "bob",
  "upperName": "BOB",
  "pascalName": "Bob",
  "evmChainId": 60808,
  "nativeSymbol": "ETH",
  "nativeName": "Ether",
  "nativePrecision": 18,
  "rpcUrl": "https://rpc.gobob.xyz",
  "explorerUrl": "https://explorer.gobob.xyz",
  "color": "#F25D00",
  "viemChainName": "bob",
  "coingeckoPlatform": "bob-network",
  "swappers": { "relay": { "chainId": 60808 } }
}
```

## what the generator handles (~80%)

All mechanical insertions:
- CAIP constants (chainId, assetId, CHAIN_REFERENCE, ASSET_REFERENCE, VALID_CHAIN_IDS)
- KnownChainIds enum + EvmChainId union
- Chain adapter class (file template) + exports
- Discriminated unions in chain-adapters/types.ts (Account, FeeData, SignTx, etc.)
- Plugin file (template) + registration in activePlugins.ts
- Feature flags (preferencesSlice, config.ts, store mock, vite-env.d.ts)
- .env / .env.development entries
- Viem client creation + all 3 maps
- Ethers provider switch case
- Base asset definition
- All 5 utility functions (getBaseAsset, chainIdToFeeAssetId, etc.)
- CSP headers (template + registration)
- Asset generation script (template + registration)
- Coingecko adapter (enum, switch cases, parseData)
- Transaction status utility (template)
- State migration (bump version, clearAssets entry)
- HDWallet packages (all 10+ — supports{Chain} pattern)
- WalletConnect V2 (all 4 touchpoints)
- PluginProvider, MarketsRow, AssetService feature flag filters
- Portfolio utils, wallet support hook, EVM account derivation
- SECOND_CLASS_CHAINS array, popular assets, opportunities mappings
- Coingecko supported chain IDs

## what claude handles (~20%)

- **Swapper research**: check Relay, Across, Portals, Zerion, Yield.xyz docs for chain support
- **Related asset index**: look up native USDC/USDT/DAI addresses, manual mappings
- **Edge cases**: ERC20 native duplicate blacklisting, market service test counts, icon URL validation
- **Data generation**: run `yarn generate:chain`
- **Verification**: lint, type-check, runtime testing

## the workflow end-to-end

```
1. Fill chain-config.json (Claude can help research fields)
2. yarn ts-node scripts/addChain/index.ts configs/<chain>.json  (~5 seconds)
3. Claude handles research items (swappers, related assets, edge cases)
4. yarn generate:chain eip155:<chainId>
5. yarn lint --fix && yarn type-check
6. Runtime verification
```

## one contract, two modes

The existing contract stays as the **single source of truth** for both building AND reviewing chain PRs. Each item gets annotated:

```markdown
1. **CAIP Constants** `[auto]` - packages/caip/src/constants.ts
   - <chain>AssetId, <chain>ChainId, ...

27. **Relay Swapper** `[research]` - packages/swapper/src/swappers/RelaySwapper/constant.ts
    - Check https://relay.link for supported chains
```

Then:
- **Build mode** (`/chain-integration` skill): skips `[auto]` items (generator handles them), only walks through `[research]` items
- **Review mode** (PR review skill): reads ALL items regardless of tag, does full gap analysis against the PR diff — if an `[auto]` item is missing, the generator wasn't used or has a bug, either way the review catches it

The contract never shrinks. The tags just tell the build skill "don't waste brain on this, the generator already did it."

## implementation order

1. **Add markers** — place ~60 `// @chain-gen:*` comments across existing files (purely additive, zero behavior change)
2. **Build generator** — schema, templates, snippet registry, engine, report
3. **Test roundtrip** — delete an existing chain (Bob), re-generate, diff against original
4. **Annotate contract** — add `[auto]`/`[research]` tags to every item
5. **Update `/chain-integration` skill** — run generator first, then research items only
6. **CI integration** — marker verification job (fails if any marker deleted during refactors)

## references

- Current contract: `.claude/contracts/second-class-evm-chain.md`
- Full plan with detailed marker placement map: `.claude/plans/atomic-wibbling-stream.md`

### PR DESCRIPTION
**Pull request by gomesalexandre**
## Description

Two things in one:

1. **Abstract chain** — adds Abstract (EVM, chain ID 2741, Pudgy Penguins' L2 on ZK Stack) as a second-class EVM chain behind `VITE_FEATURE_ABSTRACT`. Native asset is ETH, swapper support via Relay (~1-2s bridge and on-chain swap times). Tested end-to-end: bridge in, on-chain DEX swap (ETH→USDT on Abstract), notification center, balance polling.

2. **`scripts/addChain/` scaffolder** — automates ~80% of the second-class EVM chain integration points documented in `.claude/contracts/second-class-evm-chain.md`. Uses the TypeScript compiler API (zero new deps — `typescript` already installed) to locate insertion points via AST, then string-splices. Abstract is the reference config (`chains/abstract.json`).

Usage: `npx tsx scripts/addChain/index.ts chains/<chain>.json`

Scaffolder covers ~50 insertion points:
- caip constants, KnownChainIds enum, chain adapter types
- hdwallet support flags across all wallet implementations
- viem client, fallback RPCs, ethers provider switch case
- feature flags, CSP headers, plugin registration
- Relay swapper chain mapping + relayTokenToAssetId switch
- asset generator, coingecko adapter, popular assets, markets row
- WalletConnect config, account derivation, portfolio utils

Append-only ordering via `appendLineAfterLastPattern` so running the scaffolder for chain B always inserts after chain A — no merge conflicts. Idempotent: safe to run twice.

## Issue (if applicable)

closes #11980

## Risk

Low — new chain behind feature flag, scaffolder is scripts-only (no runtime impact).

> No new on-chain transactions introduced. Scaffolder runs at dev-time only.

### Engineering

1. Enable `VITE_FEATURE_ABSTRACT=true` in `.env.development`
2. Start dev server: `pnpm run dev`
3. Trade page → chain filter → Abstract appears
4. ETH on Abstract shows as popular asset when filtered
5. Bridge ETH → Abstract via Relay, confirm in ~2s
6. On-Abstract DEX swap (ETH → USDT) via Relay, confirm in ~1s
7. Notification center shows "Confirmed" for both

Scaffolder test: `npx tsx scripts/addChain/index.ts chains/abstract.json` (idempotent — no-ops on already-applied entries)

### Operations

- [x] :checkered_flag: My feature is behind a flag and doesn't require operations testing (yet)

## Screenshots (if applicable)

QA run (all 21 steps passed): https://qabot-kappa.vercel.app/runs/52fafd4b-483d-426f-aaa5-841c48c690cb

## Summary by CodeRabbit

* **New Features**
  * Adds support for the Abstract (ABS) network: wallet support, asset discovery, portfolio/markets UI, trading (including Relay swap routes) and a new plugin.
  * Adds a feature flag and node URL configuration to enable/disable Abstract.

* **Chores**
  * Migration updates and asset/data wiring to include Abstract.
  * Adds CLI tooling and templates to scaffold new chain integrations and generate asset data.

* **Tests**
  * New end-to-end fixture and unit test updates covering Abstract scenarios.

### CODE REVIEW

**Comment by gomesalexandre** [comment]
@coderabbitai review

**Comment by gomesalexandre** [comment]
ser you're drunk @coderabbitai

`abstract` follows the same pattern as `ink`, `flowEvmChain`, and `sei` — all feature-flagged chains that are unconditionally included in `walletConnectV2OptionalChains` and `rpcMap`. None of them are gated by their feature flag in this file. Abstract is consistent with the existing convention.

**Comment by gomesalexandre** [comment]
done @coderabbitai — addressed in the follow-up commit:
- AbstractMainnet added to ChainSignTx ✓
- chainId 2741 added to ethSupportsNetwork() ✓  
- process.exit(1) for scaffolder errors ✓
- wrappedNativeAddress empty-string guard + relay relayChainId validation ✓
- pnpm exec tsx for add-chain script ✓
- more fallback RPCs for Abstract ✓
- coingecko utils.test.ts parseData updated ✓ (was the failing CI check too)

dismissing the wrap()/engine idempotency/evm.ts prefetch comments — wrap() already logs via console.warn on misses (surfaced to the user), addNamedImport idempotency key format is intentional, and the evm.ts prefetch uses ethGetAddresses which works for all EVM chains regardless of supportsAbstract

**Comment by NeOMakinG** [comment]
## 🤖 QA Automation Report

**QABot Run:** [#59223faa](https://qabot-kappa.vercel.app/runs/59223faa-cc5b-4b59-ae01-0df23e36f2b6)
**Status:** ⚠️ Partial Pass

### ✅ Build Verification
- `pnpm install && pnpm build:packages` succeeded
- Dev server starts on port 4471
- `VITE_FEATURE_ABSTRACT=true` enabled in .env.development
- No TypeScript errors after package build

### ⚠️ UI Testing (Partial)
- Home page loads correctly
- Trade UI renders with ETH/BTC default pair
- Search modal opens and displays assets
- **Note:** Browser automation encountered timing issues preventing full Abstract chain verification

### 📋 Recommendations
Manual verification needed for:
1. Abstract chain appears in asset selector when feature flag enabled
2. Bridge to Abstract chain works via Relay
3. On-chain swap (ETH→USDT) on Abstract
4. Balance polling and notification center

---
*Automated QA by [QABot](https://qabot-kappa.vercel.app)*

**Comment by gomesalexandre** [comment]
addressing latest @coderabbitai comments:

**ethSupportsNetwork (walletconnectV2.ts)** — done in prior commit, 2741 is already at line 139

**process.exit(1) (report.ts)** — done in prior commit

**schema.ts relay + wrappedNativeAddress guards** — done in prior commit

**shebang (index.ts)** — fixed to \`#!/usr/bin/env -S npx tsx\` in latest commit

**batch-prefetch ordering (evm.ts)** — the prefetch gates on \`wallet.ethGetAddresses\` which works for any EVM-capable wallet regardless of \`supportsAbstract\`. All other second-class EVM chains use the identical pattern — dismissing.

**wrap() masking anchor misses** — post-MVP. \`false\` from engine = skip, exception = error. The "anchor not found" vs "already present" distinction is a UX improvement but doesn't affect scaffold correctness — dismissing.

**addNamedImport() idempotency** — post-MVP, catalogued for a follow-up hardening pass.

**migration key from highest entry** — valid, post-MVP improvement.

**Across/Portals/Zerion supported branches** — acknowledged, full integration per-swapper varies. Post-MVP when those swappers get Abstract support.

**partial adapter generation idempotency** — post-MVP.

File: packages/chain-adapters/src/types.ts [hunk 1]
```diff
@@ -327,6 +329,7 @@ export type ChainSpecificBuildTxData<T> = ChainSpecific<
     [KnownChainIds.ModeMainnet]: evm.BuildTxInput
     [KnownChainIds.SoneiumMainnet]: evm.BuildTxInput
     [KnownChainIds.SeiMainnet]: evm.BuildTxInput
+    [KnownChainIds.AbstractMainnet]: evm.BuildTxInput
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply packages/chain-adapters/src/types.ts:333]: done - AbstractMainnet added to ChainSignTx type map in prior commit

File: packages/hdwallet-walletconnectv2/src/walletconnectV2.ts [hunk 2]
```diff
@@ -225,6 +225,7 @@ export class WalletConnectV2HDWallet implements HDWallet, ETHWallet, BTCWallet {
   readonly _supportsPlume = true
   readonly _supportsZkSyncEra = true
   readonly _supportsBlast = true
+  readonly _supportsAbstract = true
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply packages/hdwallet-walletconnectv2/src/walletconnectV2.ts:228]: done - 2741 is in ethSupportsNetwork() at line 139 now

File: scripts/addChain/codemods.ts [hunk 3]
```diff
@@ -0,0 +1,1587 @@
...
+  return path.join(ROOT, ...parts)
+}
+
+function wrap(file: string, operation: string, fn: () => boolean): CodemodResult {
+  try {
+    const changed = fn()
+    return { file, operation, status: changed ? 'inserted' : 'skipped' }
+  } catch (err) {
+    return { file, operation, status: 'error', message: String(err) }
+  }
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply scripts/addChain/codemods.ts:30]: post-MVP — wrap() propagates false/exception correctly; anchor-not-found vs already-present distinction is a UX improvement for the scaffold operator, not a correctness issue

File: scripts/addChain/engine.ts [hunk 4]
```diff
@@ -0,0 +1,401 @@
...
+      if (ts.isStringLiteral(moduleSpec) && moduleSpec.text === modulePath) {
+        targetImport = node
+      }
+    }
+  })
+
+  if (!targetImport) {
+    // No existing import for this module - create a new one
+    const newImport = `import { ${importName} } from '${modulePath}'\n`
+    return appendAfterLastImport(filePath, newImport, importName)
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply scripts/addChain/engine.ts:179]: post-MVP — current key covers the vast majority of real cases; catalogued for a follow-up hardening pass

File: scripts/addChain/report.ts [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+    console.log(`  [ ] ${i + 1}. ${manualTasks[i]}`)
+  }
+
+  console.log()
+
+  if (errors.length > 0) {
+    console.log(`!! ${errors.length} error(s) detected - review and fix before proceeding.\n`)
+  } else {
+    console.log('All codemods applied successfully.\n')
+  }
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply scripts/addChain/report.ts:84]: done - process.exit(1) added in prior commit

File: scripts/addChain/schema.ts [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+    }
+  }
+
+  if (c.wrappedNativeAddress !== null && typeof c.wrappedNativeAddress !== 'string') {
+    throw new Error(
+      `"wrappedNativeAddress" must be a string or null, got: ${JSON.stringify(
+        c.wrappedNativeAddress,
+      )}`,
+    )
+  }
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply scripts/addChain/schema.ts:99]: done - both guards added: wrappedNativeAddress empty-string throws, relay.relayChainId required when supported === true

File: src/lib/account/evm.ts [hunk 7]
```diff
@@ -181,6 +183,7 @@ export const deriveEvmAccountIdsAndMetadata: DeriveAccountIdsAndMetadata = async
     if (chainId === storyChainId && !supportsStory(wallet)) continue
     if (chainId === zkSyncEraChainId && !supportsZkSyncEra(wallet)) continue
     if (chainId === blastChainId && !supportsBlast(wallet)) continue
+    if (chainId === abstractChainId && !supportsAbstract(wallet)) continue
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply src/lib/account/evm.ts:186]: dismissing — the prefetch gates on wallet.ethGetAddresses (any EVM wallet), not supportsAbstract. all other second-class EVM chains use the identical pattern without issue

File: scripts/addChain/codemods.ts [hunk 8]
```diff
@@ -0,0 +1,1615 @@
...
+      const lastNum = Math.max(...matches.map(m => parseInt(m[1], 10)))
+      const nextNum = lastNum + 1
+      if (src.includes(`${nextNum}: clearAssets`)) return false
+      return appendLineAfterPattern(
+        migrationsFile,
+        new RegExp(`${lastNum}: clearAssets,`),
+        `  ${nextNum}: clearAssets,`,
+        `${nextNum}: clearAssets`,
+      )
+    }),
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply scripts/addChain/codemods.ts:1033]: post-MVP improvement — noted

File: scripts/addChain/codemods.ts [hunk 9]
```diff
@@ -0,0 +1,1615 @@
...
+  if (c.swappers.zerion.supported) {
+    const zerionFile = r('packages/types/src/zerion.ts')
+    if (fs.existsSync(zerionFile)) {
+      results.push(
+        wrap(zerionFile, 'zerion-chain-support', () =>
+          addNamedImport(zerionFile, '@shapeshiftoss/caip', `${c.camelName}ChainId`),
+        ),
+      )
+    }
+  }
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply scripts/addChain/codemods.ts:1135]: acknowledged — full swapper integration varies per swapper; will tighten when those swappers get Abstract support

File: scripts/addChain/codemods.ts [hunk 10]
```diff
@@ -0,0 +1,1615 @@
...
+      status: 'inserted',
+    })
+  } else {
+    results.push({
+      file: adapterFile,
+      operation: 'create-chain-adapter',
+      status: 'skipped',
+      message: 'already exists',
+    })
+  }
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply scripts/addChain/codemods.ts:1552]: post-MVP — noted

File: scripts/addChain/index.ts [hunk 11]
```diff
@@ -0,0 +1,32 @@
+#!/usr/bin/env npx tsx
```
[the remark this answers is not in the document]
Comment (gomesalexandre) [reply scripts/addChain/index.ts]: done - fixed to #!/usr/bin/env -S npx tsx in latest commit
